### PR TITLE
Introduce Version property to messages

### PR DESCRIPTION
**Pull request by limebell**
Removed `Peer.AppProtocolVersion` which were used to check version between swarms. Instead, added `AppProtocolVersion`-typed `Version` property to `Message` and use it to check validity of messages.

If received message has different app protocol version, `NetMQTransport` will reply a `DifferentVersion` message regardless of received message's type and does not process anything.

### CODE REVIEW

File: Libplanet/Net/Peer.cs [hunk 1]
```diff
@@ -90,53 +61,10 @@ protected Peer(SerializationInfo info, StreamingContext context)
...
-        /// to trust no one, so that this method always returns <c>false</c>.</em>  Conversely,
-        /// <c>null</c> means to <em>trust anyone</em>.  Be aware of the difference of those two.
-        /// </param>
-        /// <param name="recognizer">A callback to determine an encountered peer's
-        /// <see cref="AppProtocolVersion"/> is compatible with the currently running
-        /// application, in the case of the peer's version signature is trustworthy but its
-        /// version number is not the same.</param>
-        /// <returns><c>true</c> if the peer is compatible with the given criteria.
-        /// Otherwise <c>false</c>.</returns>
-        public bool IsCompatibleWith(
```
Comment (earlbread) [inline Libplanet/Net/Peer.cs:107]: This should be added to the changelog.

File: Libplanet/Net/Peer.cs [hunk 2]
```diff
@@ -67,14 +46,6 @@ protected Peer(SerializationInfo info, StreamingContext context)
         [Pure]
         public PublicKey PublicKey { get; }
 
-        /// <summary>
-        /// The corresponding application protocol version of this peer.
-        /// </summary>
-        // FIXME: This would be better to be nullable...
-        [IgnoreDuringEquals]
-        [Pure]
-        public AppProtocolVersion AppProtocolVersion { get; }
```
Comment (earlbread) [inline Libplanet/Net/Peer.cs:76]: This should be added to the changelog.

File: Libplanet/Net/Peer.cs [hunk 3]
```diff
@@ -20,39 +15,23 @@ namespace Libplanet.Net
     [Equals]
     public class Peer : ISerializable
     {
-        private static readonly Codec _codec = new Codec();
-
-        public Peer(
-            PublicKey publicKey,
-            AppProtocolVersion appProtocolVersion)
```
Comment (earlbread) [inline Libplanet/Net/Peer.cs:27]: This should be added to the changelog.

File: Libplanet/Net/NetMQTransport.cs [hunk 4]
```diff
@@ -947,19 +956,29 @@ private async Task InitializeTurnClient()
...
+                    return false;
+                }
+                else
+                {
+                    return _differentAppProtocolVersionEncountered(
+                        message.Remote, remoteVersion, _appProtocolVersion);
+                }
             }
+
+            return false;
```
Comment (earlbread) [inline Libplanet/Net/NetMQTransport.cs:981]: How about this?

```suggestion
            if (!(_trustedAppProtocolVersionSigners is null) &&
                !_trustedAppProtocolVersionSigners.Any(s => remoteVersion.Verify(s)))
            {
                return false;
            }

            if (_differentAppProtocolVersionEncountered is null)
            {
                return false;
            }

            return _differentAppProtocolVersionEncountered(
                message.Remote, remoteVersion, _appProtocolVersion);
```

File: Libplanet/Net/BoundPeer.cs [hunk 5]
```diff
@@ -17,22 +17,18 @@ public sealed class BoundPeer : Peer
         /// <see cref="Peer"/>.</param>
         /// <param name="endPoint">A <see cref="DnsEndPoint"/> consisting of the
         /// host and port of the <see cref="Peer"/>.</param>
-        /// <param name="appProtocolVersion">An application protocol version
-        /// that the <see cref="Peer"/> is using.</param>
         public BoundPeer(
             PublicKey publicKey,
-            DnsEndPoint endPoint,
-            AppProtocolVersion appProtocolVersion)
```
Comment (earlbread) [inline Libplanet/Net/BoundPeer.cs:25]: This should be added to the changelog.

File: Libplanet/Net/NetMQTransport.cs [hunk 6]
```diff
@@ -947,19 +956,27 @@ private async Task InitializeTurnClient()
             }
         }
 
-        // FIXME: This method should be in Swarm<T>
-        private void ValidateSender(Peer peer)
+        private bool IsMessageValid(Message message)
```
[the remark this answers is not in the document]
Comment (limebell) [reply Libplanet/Net/NetMQTransport.cs:959]: According to [dotnet naming guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members#names-of-properties), the member with boolean type is recommended to name that can be answered with "yes" or "no". So how about name it as `IsAppProtocolVersionValid()`?

[the remark this answers is not in the document]
Comment (longfin) [reply Libplanet/Net/NetMQTransport.cs:959]: That looks also good. 👍

**Comment by limebell** [comment]
I moved version check progress to `Message.Parse()`, which will throw `DifferentAppProtocolVersionException` even the frame is changed a lot.

File: Libplanet.Tests/Libplanet.Tests.csproj [hunk 7]
```diff
@@ -18,13 +16,11 @@
     <AdditionalFiles Include="..\stylecop.json" />
   </ItemGroup>
 
-  <PropertyGroup Condition=" '$(MSBuildRuntimeType)'!='Mono' And
-                             '$(MSBuildVersion)'&gt;='16.3.0'">
+  <PropertyGroup Condition=" '$(MSBuildRuntimeType)'!='Mono' And&#xD;&#xA;                             '$(MSBuildVersion)'&gt;='16.3.0'">
```
Comment (longfin) [inline Libplanet.Tests/Libplanet.Tests.csproj:19]: is this change needed?

File: Libplanet/Net/DifferentAppProtocolVersionException.cs [hunk 8]
```diff
@@ -24,6 +27,7 @@ internal sealed class DifferentAppProtocolVersionException : Exception
         /// </param>
         public DifferentAppProtocolVersionException(
             string message,
+            byte[] identity,
```
Comment (longfin) [inline Libplanet/Net/DifferentAppProtocolVersionException.cs:30]: I guess `identity` isn't useful because it's a temporary value for `NetMQRouter`.

Comment (limebell) [reply Libplanet/Net/DifferentAppProtocolVersionException.cs:30]: I've added this parameter just for use in [here](https://github.com/planetarium/libplanet/pull/949/files#diff-9045fd925fccb8bca88db49a0ab95d0cR600-R608).

File: Libplanet.Tests/Libplanet.Tests.csproj (hunk 7 above)
Comment (limebell) [reply Libplanet.Tests/Libplanet.Tests.csproj:19]: Added by mistake 😥. I'll fix it.

File: Libplanet/Net/DifferentAppProtocolVersionException.cs (hunk 8 above)
Comment (longfin) [reply Libplanet/Net/DifferentAppProtocolVersionException.cs:30]: 🆗

File: Libplanet/Net/Messages/Message.cs [hunk 9]
```diff
@@ -94,26 +97,67 @@ internal enum MessageType : byte
...
                 throw new ArgumentException("Can't parse empty NetMQMessage.");
             }
 
-            // (reply == true)  [type, sign, peer, frames...]
-            // (reply == false) [identity, type, sign, peer, frames...]
-            int headerCount = reply ? 3 : 4;
+            // (reply == true)  [version, type, sign, peer, frames...]
+            // (reply == false) [identity, version, type, sign, peer, frames...]
+            int headerCount = reply ? CommonFrames : CommonFrames + 1;
+            var versionToken = raw[headerCount - 4].ConvertToString();
```
Comment (riemannulus) [inline Libplanet/Net/Messages/Message.cs:134]: IMO If we use a constant here, it will be easy to make mistakes. if changed `CommonFrames`, easily forgot about change const number. Can we change like that?

```suggestion
            var versionToken = raw[headerCount - CommonFrames].ConvertToString();
```
or something more semantic variable name.

Comment (limebell) [reply Libplanet/Net/Messages/Message.cs:134]: Then these lines:

https://github.com/planetarium/libplanet/blob/d46118902d7876c7ea20f25c7eedccf33677d9fa/Libplanet/Net/Messages/Message.cs#L161-L163

should be like this as well?

```
 var rawType = (MessageType)raw[headerCount - (CommonFrames - 1)].ConvertToInt32(); 
 var peer = raw[headerCount - (CommonFrames - 2)].ToByteArray(); 
 byte[] signature = raw[headerCount - (CommonFrames - 3)].ToByteArray(); 
```

Comment (riemannulus) [reply Libplanet/Net/Messages/Message.cs:134]: I think it seems OK. any opinion? @libplanet

Comment (longfin) [reply Libplanet/Net/Messages/Message.cs:134]: Or, it would be more readable if we slice the `raw` array after checking 'reply'.

```charp
byte[] remains  = (reply) ? raw.Skip(1).ToArray() : raw;

// using remains instead of raw
```

Comment (limebell) [reply Libplanet/Net/Messages/Message.cs]: I've added `MessageFrame` enum in [`e8eacb83f1d6f9a96d363f791b38752bc781884d`](https://github.com/planetarium/libplanet/pull/949/files#diff-4008fb0720717f5ed3e957339adebf4dR120).
